src/compiler/checker.ts

```diff
@@ -5912,10 +5913,11 @@ module ts {
                     // No need to special-case tagged templates; their excludeArgument value will be 'undefined'.
                     if (excludeArgument[i] === false) {
                         var parameterType = getTypeAtPosition(signature, i);
-                        inferTypes(context, checkExpressionWithContextualType(args[i], parameterType, mapper), parameterType);
+                        inferTypes(context, checkExpressionWithContextualType(args[i], parameterType, inferenceMapper), parameterType);

```

Comment:
So the order is essentially still the same. We first do all the arguments (instead of just the ones that were never excluded). Then we come back and do the ones that were once excluded but no longer are.

I'm noticing two things:

1. For the arguments that do contain context sensitive expressions, we will end up visiting (inferring from) most of the argument twice. The only parts that will be new the second time are the context sensitive expressions. I guess it is not worth the complexity of trying to fish out just the context sensitive expressions for this second visit. After all, many arguments of this sort are likely to be lambdas, which only consist of a context sensitive expression.
2. We will visit each argument at most twice. So while this does improve things over what we had before, we will not be able to handle something like:
```ts
interface Computed<T, U> {
    read(): T;
    write(value: T): U;
    rewrite(value: U);
}

function foo<T, U>(x: Computed<T, U>) { }

var s: string;

foo({
    read: () => s,
    write: value => s = value,
    rewrite: value => s = value
});
```

For this to work, you'd have to visit the argument 3 times. I would be very surprised if this kind of thing ever comes up though.